crates/core/c8y_api/src/http_proxy.rs

```diff
@@ -69,6 +73,14 @@ pub trait C8YHttpProxy: Send + Sync {
         config_type: &str,
         child_device_id: Option<String>,
     ) -> Result<String, SMCumulocityMapperError>;
+
+    async fn download_file(
+        &mut self,
+        download_url: &str,
+        file_name: &str,
+        version: &Option<String>,
+        tmp_dir_path: &Path,
+    ) -> Result<PathBuf, SMCumulocityMapperError>;
```

Comment:
Looking at the code of the `Downloader`: what is proposed by @albinsuresh can easily be done in this PR.

* [The `file_name`, `version` and `tmp_dir_path` are  *only* used to forge the download target file path](https://github.com/thin-edge/thin-edge.io/blob/main/crates/common/download/src/download.rs#L72).
* The logic to forge a path from package name and version should be moved in `crates/core/plugin_sm/src/plugin.rs`. This is the only place were a version is provided ;-).
* The `Downloader` constructor should *only* take a file path, the full path to where we expect the file to be downloaded. 
* By the way, `tmp_dir_path` is a misnomer. The download is actually down there and the `Download` struct doesn't move any file from a tmp to a target location.
* Since downloading first in a temporary dir is good practice, I would implement the move on success logic here.

So, I would simplify this method to 2 arguments, possibly 3.

```suggestion
    async fn download_file(
        &mut self,
        download_url: &str,       // source url
        target_file: &Path,       // full path to the target file
        temp_dir: Option<&Path>, // If provided, the file is downloaded there first and move to the target on success.
    ) -> Result<PathBuf, SMCumulocityMapperError>;
```

Reply:
> Taking arguments like version is too specific to our software mgmt agent. I'd keep this very generic with just the file name and temp directory. But, I understand that this field requirement is coming from the Download API, so, leave it as-is for now.

As you mentioned, the download API requires a version. If it is given, the downloaded file name will be `<filename>_<version>`. If not, just `<filename>`. But actually I found the version is usual also for firmware plugin. So, it makes sense to keep version as an argument for me (also it's already an `Option`)

> Ideally, I'd have kept the file name and temp dir also optional. If the temp directory is not provided, we could default to some temp directory (e.g. /tmp). If the dest_path is not provided, we could download it to some temp location and return that path.

This can be a good enhancement. I think you can do it later in your refactoring :D

Reply:
I did an middle solution.

The `download_file` API in http_proxy crate has now more flexibility on the file name. If a software name `curl` and version `1.0.0` and want to have a file name `curl_1.0.0`, then now you just need to give `curl_1.0.0` as `file_name`.
```rust
    async fn download_file(
        &mut self,
        download_url: &str,
        file_name: &str,
        tmp_dir_path: &Path,
    ) -> Result<PathBuf, SMCumulocityMapperError>;
```

To do so, 
* renamed `downloader::new()` to `downloader::new_sm()`, so that I don't need to change the current code so much.
* added `downloader_new()`, which takes two arguments only: `file_name` and `tmp_dir`.

However, the `downloader::new_sm()` is a temporary solution. It should be removed and use only `downloader::new()` in the future. I guess @albinsuresh can do it when he does further refactoring.